dual_net.py

```diff
@@ -157,13 +103,93 @@ def __init__(self,
         self.update_op = tf.train.AdamOptimizer(learning_rate).minimize(self.loss)
         self.sess = sess
 
+
+    def build_model(self,
+                    board_placeholder,
+                    legality_mask_placeholder,
+                    scope,
+                    shared_layers,
+                    policy_head,
+                    value_head):
+        """
+        Returns the output tensors for an model based on the layers in shared_layers,
+        policy_head, and value_head.
+        shared_layers is a list of dicts, each dict representing a layer.
+        - Convolutional layers:
+           - d['layer'] <- 'conv'
+           - d['num_outputs'], d['kernel_size'], d['stride'] should be ints
+           - d['activation_fn'] is a function
+        - Residual layers:
+           - d['layer'] <- 'residual'
+           - other keys same as convolutional
+        - Fully connected layers:
+           - d['layer'] <- 'fc'
+           - d['num_outputs'] is an int
+           - d['activation_fn'] is a function
+        policy_head and value_head have the same structure as above but represent
+        the layers for the policy head and value head, respectively.
+
+        returns the policy output and the value output in a tuple
+        """
+        out = board_placeholder
+        for specs in shared_layers:
+            if specs['layer'] == 'conv':
+                out = conv_block(out, specs)
+            elif specs['layer'] == 'residual':
+                out = residual_block(out, specs)
+            elif specs['layer'] == 'fc':
+                out = layers.flatten(out)
+                out = layers.fully_connected(out,
+                                             num_outputs=specs['num_outputs'],
+                                             activation_fn=specs['activation_fn'])
+        # Policy head
+        policy_out = out
+        for specs in policy_head:
+            if specs['layer'] == 'conv':
+                policy_out = conv_block(policy_out, specs)
+            elif specs['layer'] == 'fc':
+                policy_out = layers.flatten(policy_out)
+                policy_out = layers.fully_connected(policy_out,
+                                                    num_outputs=specs['num_outputs'],
+                                                    activation_fn=specs['activation_fn'])
+
+        x = tf.exp(policy_out) * legality_mask_placeholder
+        # Needs reshape to broadcast properly 
+        policy_out = x / tf.reshape(tf.reduce_sum(x, axis=1), shape=((tf.shape(x)[0],) + (1,)))
```

Comment:
I'm a bit confused as to why we have to do this reshape alchemy..

```
(Pdb) x / tf.reshape(tf.reduce_sum(x, axis=1), shape=((tf.shape(x)[0],) + (1,)))
<tf.Tensor 'truediv:0' shape=(?, 4096) dtype=float32>
(Pdb) x / (tf.reduce_sum(x, axis=1))
<tf.Tensor 'truediv_1:0' shape=(?, 4096) dtype=float32>
```

Looks like that second expression gives us the shape we want. What am I missing?

Reply:
Just tried again, when I do the second expression I get:
InvalidArgumentError (see above for traceback): Incompatible shapes: [10,4096] vs. [10]